### PR TITLE
OverflowMode, enablePulsingAnimation rework & other updates

### LINKED ISSUE #33: enablePulsingAnimation does not support Hot Reload
When a DescribedFeatureOverlay is already showing the overlay, changing the `enablePulsingAnimation` parameter and then hot reloading will not change the state of the pulsing animation.

It should *stop* the pulsing animation (reset to 0, not pause in the middle of the animation) when changing from `false` to `true` and *start* it in the other case.

---

This seems like a dev improvement at first, but having the ability to e.g. start the pulsing animation while already being in feature discovery programmatically can be quite useful (when rebuilding a widget and changing the `enablePulsingAnimation` parameter dynamically).

## Current situation

Currently, the widget needs to be reinserted into the tree completely.

### LINKED ISSUE #34: Handle content that overflows the overlay
Currently, the `title` and `description` `Text`s can easily overflow the overlay making the overlay both ugly and unreadable.

The reason this should be handled in the package and not by the user is that the space e.g. `Text`s take up also depends on the system font size, which makes it hard to control the allowed dimension from outside of the package.

### PR DESCRIPTION
**Pull request by creativecreatorormaybenot**
Resolves #33 and resolves #34.

### CODE REVIEW

**Comment by creativecreatorormaybenot** [comment]
@axel-op Pulsing is done I think. You can already review that if you like. I will do the overflow tomorrow.

**Comment by creativecreatorormaybenot** [comment]
@axel-op I got `OverflowMode.clip` working now. What do you think about the default? I think `OverflowMode.clip` is a nicer default in terms of design because it looks really ugly when text overflows the background, but on the other hand, these overlays are meant to explain features and it is crucial that all the text is visible - you cannot be sure what the font size is going to be.

Regarding `OverflowMode.cover`, I am not exactly sure how to implement that yet. The part for figuring out if the content area is inside of the background area is trivial (just some trigonometry), but I need to do some research for finding out what the area of the content is going to be. Ideally, I would catch that during layout, which might require custom `RenderObject`s.

I might add some tests for `clip` first, I will see.

**Comment by creativecreatorormaybenot** [comment]
This was so much harder than I thought it would be (and that was only clipping).

File: lib/src/described_feature_overlay.dart [hunk 1]
```diff
@@ -806,49 +821,164 @@ class _Content extends StatelessWidget {
...
+                            ]))))));
+
+    if (overflowMode == OverflowMode.clip)
+      result = _ClipContent(
+        backgroundPosition: backgroundPosition,
+        backgroundRadius: backgroundRadius,
+        child: result,
+      );
+
+    result = Positioned(
```
Comment (axel-op) [inline lib/src/described_feature_overlay.dart]: @creativecreatorormaybenot Shouldn't you add an `else` there?
EDIT: Oh no OK I see

File: lib/src/described_feature_overlay.dart [hunk 2]
```diff
@@ -663,33 +680,33 @@ class _Content extends StatelessWidget {
...
+    this.touchTargetRadius,
+    this.title,
+    this.description,
+    this.state,
+    this.transitionProgress,
+    this.orientation,
+    this.textColor,
+    this.overflowMode,
+    this.backgroundRadius,
+    this.backgroundPosition,
```
Comment (axel-op) [inline lib/src/described_feature_overlay.dart]: Why did you remove the `@required` annotations?

File: lib/src/described_feature_overlay.dart [hunk 3]
```diff
@@ -143,10 +147,23 @@ class _DescribedFeatureOverlayState extends State<DescribedFeatureOverlay>
...
     super.dispose();
   }
 
+  @override
+  void didUpdateWidget(DescribedFeatureOverlay oldWidget) {
+    if (oldWidget.enablePulsingAnimation != widget.enablePulsingAnimation) {
+      if (widget.enablePulsingAnimation)
+        _pulseController.forward(from: 0);
+      else {
+        _pulseController.stop();
```
Comment (axel-op) [inline lib/src/described_feature_overlay.dart:166]: May I suggest to add a listener (removed when `enablePulsingAnimation` is true) that stops the controller only when the animation is completed?

**Comment by creativecreatorormaybenot** [comment]
@axel-op I will call the previous `doNothing` mode `ignore`, is that fine?

Reasoining is that I want it to really show that the package does not nothing in that case, like it is now.

Btw, the reason `CHANGELOG.md` is formatted to a particular line length is that IDEs will not wrap lines, so you would need to scroll horizontally, however, adding line breaks makes it readable in IDEs and Markdown does not care about it as long as there are no double trailing spaces.

**Comment by creativecreatorormaybenot** [comment]
@axel-op I restructed the Dart files of the package.

I tried to imitate the structure the main Flutter package uses. https://github.com/flutter/flutter/tree/master/packages/flutter/lib

Let me know what you think.  
The only thing I did different is that I moved the library files that `export` their sub files into the `src` folder because they should not appear for users in code completion.

**Comment by creativecreatorormaybenot** [comment]
I implemented an attempt of making `OverflowMode.extendBackground` work. I am not sure about the calcuation for the size, but it works for now (test feature in the example). I think what I would really want to do is find the point in the `_Content` that is furthest away from the `backgroundCenter` and calculate the distance to that to make that the radius and maybe add a small padding to it (I would include that padding in the `extendBackground` documentation).

@axel-op Do you have an idea of how to do that? I will add tests first, I think.  
I am fairly confident about my `src/rendering` implementations. Not sure how familiar you are with the rendering layer in Flutter, but you can take a look if you want.

File: pubspec.yaml [hunk 4]
```diff
@@ -3,15 +3,15 @@ description: >-
   A Flutter package that implements Material Design Feature discovery
   to show a description of specific features to new users.
   See https://tinyurl.com/FeatureDiscovery
-version: 0.5.1
+version: 0.6.0
```
Comment (axel-op) [inline pubspec.yaml]: Why did you change your mind about that?

File: lib/src/widgets/overlay.dart [hunk 5]
```diff
@@ -143,14 +154,27 @@ class _DescribedFeatureOverlayState extends State<DescribedFeatureOverlay>
...
     super.dispose();
   }
 
+  @override
+  void didUpdateWidget(DescribedFeatureOverlay oldWidget) {
+    if (oldWidget.enablePulsingAnimation != widget.enablePulsingAnimation) {
+      if (widget.enablePulsingAnimation)
+        _pulseController.forward(from: 0);
+      else {
+        _pulseController.stop();
```
Comment (axel-op) [inline lib/src/widgets/overlay.dart:166]: This stops the animation abruptly. I think you should register a callback that stops the animation once it has completed.

File: lib/src/widgets/overlay.dart [hunk 6]
```diff
@@ -143,14 +154,27 @@ class _DescribedFeatureOverlayState extends State<DescribedFeatureOverlay>
...
+      }
+    }
+    super.didUpdateWidget(oldWidget);
+  }
+
   @override
   void didChangeDependencies() {
     super.didChangeDependencies();
-    final Bloc bloc = FeatureDiscovery._blocOf(context);
+    final Bloc bloc = Provider.of<Bloc>(context, listen: false);
```
Comment (axel-op) [inline lib/src/widgets/overlay.dart]: Why? The previous way to get the bloc should be preferred as it contains an assertion that checks if there's actually a bloc

File: lib/src/widgets/overlay.dart [hunk 7]
```diff
@@ -179,12 +203,16 @@ class _DescribedFeatureOverlayState extends State<DescribedFeatureOverlay>
     });
   }
 
+  String activeFeatureId;
```
Comment (axel-op) [inline lib/src/widgets/overlay.dart]: This doesn't seem like a good idea, as this reintroduces a dependency towards the bloc's logic

File: lib/src/widgets/overlay.dart [hunk 8]
```diff
@@ -356,125 +551,37 @@ class _DescribedFeatureOverlayState extends State<DescribedFeatureOverlay>
...
-    @required this.state,
-    @required this.transitionProgress,
-    @required this.orientation,
-  })  : assert(anchor != null),
-        assert(color != null),
-        assert(screenSize != null),
+    this.color,
+    this.state,
+    this.transitionProgress,
+    this.overflowMode,
```
Comment (axel-op) [inline lib/src/widgets/overlay.dart]: Why did you remove the `@required` annotations?

File: lib/src/widgets/content.dart [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    this.title,
+    this.description,
+    this.state,
+    this.transitionProgress,
+    this.textColor,
+    this.overflowMode,
+    this.backgroundRadius,
+    this.backgroundCenter,
+    this.offsetMultiplier,
+    this.width,
```
Comment (axel-op) [inline lib/src/widgets/content.dart]: All of these should be annotated `@required`

File: lib/src/rendering/custom_layout.dart [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  final Offset contentPosition;
+
+  final Offset backgroundCenter;
+  final double backgroundRadius;
+
+  BackgroundContentLayoutDelegate({
+    this.overflowMode,
+    this.contentPosition,
+    this.backgroundCenter,
+    this.backgroundRadius,
```
Comment (axel-op) [inline lib/src/rendering/custom_layout.dart]: All of these should be `@required` too because we have to use them in our implementation!

File: lib/src/rendering/custom_layout.dart [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        assert(backgroundCenter != null),
+        assert(backgroundRadius != null);
+
+  @override
+  void performLayout(Size size) {
+    assert(hasChild(BackgroundContentLayout.background));
+    assert(hasChild(BackgroundContentLayout.content));
+
+    final contentSize =
+        layoutChild(BackgroundContentLayout.content, BoxConstraints());
```
Comment (axel-op) [inline lib/src/rendering/custom_layout.dart]: `BoxConstraints` can be `const`

**Review by axel-op — COMMENTED** [review]
See my different comments.

What's the point in regorganizing files like you did, as they won't be publicly exposed anyway?

File: lib/src/described_feature_overlay.dart (hunk 3 above)
Comment (creativecreatorormaybenot) [reply lib/src/described_feature_overlay.dart:166]: @axel-op To make it look better?

I thought about this initially, but I found three reasons why I did not do it:

 1. It does not look bad to stop the pulsing abruptly. For hot reload, it would be somewhat annoying if it did not stop immediately. If one were to stop it programmatically, they would probably have some other change in UI to accompany it.

 1. This is also how the pulsing animation is stopped when closing the overlay.

 1. It adds unnecessary weight and complexity. I cannot imagine someone using it programmatically anyway. Maybe someone has a scenario for it, but it is very helpful for debugging.

File: lib/src/widgets/overlay.dart (hunk 5 above)
Comment (creativecreatorormaybenot) [reply lib/src/widgets/overlay.dart:166]: https://github.com/axel-op/feature_discovery/pull/35#discussion_r329324166

File: lib/src/widgets/overlay.dart (hunk 6 above)
Comment (creativecreatorormaybenot) [reply lib/src/widgets/overlay.dart]: Then we should make it public. As you said, we should not use the `part` directive.

File: lib/src/described_feature_overlay.dart (hunk 3 above)
Comment (axel-op) [reply lib/src/described_feature_overlay.dart:166]: You're right, except that for your number 2, we should find a way to change that (in the dismiss method of the overlay)

File: lib/src/widgets/overlay.dart (hunk 7 above)
Comment (creativecreatorormaybenot) [reply lib/src/widgets/overlay.dart]: I did not see that `Bloc` saves this as well. Should I use that instead?

File: lib/src/rendering/custom_layout.dart (hunk 10 above)
Comment (creativecreatorormaybenot) [reply lib/src/rendering/custom_layout.dart]: I think it's a waste, but I will add them anyway... This code cannot magically change for the user 😅

File: lib/src/widgets/overlay.dart (hunk 6 above)
Comment (axel-op) [reply lib/src/widgets/overlay.dart]: Right, since you changed the organization of files.

**Comment by creativecreatorormaybenot** [comment]
@axel-op Thanks for reviewing. I will quickly implement the suggestions you already made, however, I am currently still finishing the `OverflowMode` implementation. You would need to review that again.

Why would I not organize the files? I think it is a good idea to organize files in a project. It also makes import statements in the package more concise and requires less of them.  
Btw, GitHub detects when a file has just been renamed or moved and will keep the Git history for the lines of code. You can check Git blame..

**Comment by creativecreatorormaybenot** [comment]
Btw, if you are wondering about the `meta` import. It is needed for the in my opinion unnecessary `@required` annotations.

File: lib/src/widgets/overlay.dart (hunk 6 above)
Comment (creativecreatorormaybenot) [reply lib/src/widgets/overlay.dart]: Actually, check my next commit.

File: lib/src/widgets/overlay.dart (hunk 7 above)
Comment (axel-op) [reply lib/src/widgets/overlay.dart]: I replaced this by a bool variable named `_isShowing`, tell me what you think.

**Comment by creativecreatorormaybenot** [comment]
Sorry that I removed your commit there, I should have rolled it back or merged it, but my Git skills are not great. Can you explain what it did again or push it to another branch?

File: lib/src/rendering/custom_layout.dart (hunk 10 above)
Comment (axel-op) [reply lib/src/rendering/custom_layout.dart]: It's not for the user, it's just that this is so much more convenient for debugging!

**Comment by axel-op** [comment]
> Can you please wait until I finished the implementation @axel-op?
> Sorry that I removed your commit there, I should have rolled it back, but my Git skills are not great. Can you explain what it did again or push it to another branch?

You literally pushed out my commit 😲

**Comment by creativecreatorormaybenot** [comment]
@axel-op I know, it was not on purpose. I did not realize that it would happen when I wanted to push my commit for the review comments..

Btw, does Pana say why the health score is 74.62?

**Comment by axel-op** [comment]
> Btw, does Pana say why the health score is 74.62?

It was before you removed the `_blocOf` method, because there was a problem with it ('health' errors aren't reported yet, but they will be soon). 
There is no comment when both scores are 100/100.

**Comment by creativecreatorormaybenot** [comment]
@axel-op What did you do there? I cannot see it. Can I force push?..

**Comment by creativecreatorormaybenot** [comment]
I finished the implementation. I think that I will only need to do tests now. Do you think that `outerContentPadding` is a good thing and what do you think about `6` as a value?

File: lib/src/widgets/overlay.dart [hunk 12]
```diff
@@ -656,202 +750,36 @@ class _TapTarget extends StatelessWidget {
...
+///    i.e. clip the content.
+///    Additionally, it will discard any hit events that occur outside of the
+///    inner area, so you do not have to worry about that.
+///  * [extendBackground] will expand the background circle if necessary.
+///    The radius will be increased until the content fits within the circle's area
+///    and a padding of [BackgroundContentLayoutDelegate.outerContentPadding] will be added.
+///  * [wrapBackground] does what [extendBackground] does if the content is larger than the background,
+///    but it will shrink the background if it is smaller than the content additionally.
+///    This will never be smaller than the `min(screenWidth, screenHeight) + BackgroundContentLayoutDelegate.outerContentPadding`
+///    because the furthest point of empty content will be `min(screenWidth, screenHeight)` away from the center of the overlay
```
Comment (axel-op) [inline lib/src/widgets/overlay.dart:772]: This enum is public, right? Because I don't think this will make much sense to users as they don't have access to `BackgroundContentLayoutDelegate`

**Comment by creativecreatorormaybenot** [comment]
@axel-op I think your `isShowing` implementation is not required anymore, right?

**Review by axel-op — APPROVED** [review]

**Comment by axel-op** [comment]
> @axel-op I think your isShowing implementation is not required anymore, right?

Indeed, that's why I haven't reimplemented it.

**Comment by creativecreatorormaybenot** [comment]
@axel-op Erm. what was that commit?

File: lib/src/foundation/bloc.dart [hunk 13]
```diff
@@ -3,6 +3,21 @@ import 'dart:async';
 import 'package:flutter/widgets.dart';
 import 'package:provider/provider.dart';
 
+class BlocProvider extends StatelessWidget {
```
Comment (creativecreatorormaybenot) [inline lib/src/foundation/bloc.dart:6]: Why?

Comment (axel-op) [reply lib/src/foundation/bloc.dart:6]: Because we will be sure that the `dispose` method is called as now this responsibility is only inside the bloc.dart file

File: lib/src/widgets/overlay.dart (hunk 12 above)
Comment (creativecreatorormaybenot) [reply lib/src/widgets/overlay.dart:772]: I will update it.

**Comment by axel-op** [comment]
> I finished the implementation. I think that I will only need to do tests now. Do you think that outerContentPadding is a good thing and what do you think about 6 as a value?

Maybe this should be a multiple of 4 but that's not very important. 

Anyway, could you launch the example's main file and tell me if everything is fine for you?

**Comment by axel-op** [comment]
> I implemented an attempt of making OverflowMode.extendBackground work. I am not sure about the calcuation for the size, but it works for now (test feature in the example). I think what I would really want to do is find the point in the _Content that is furthest away from the backgroundCenter and calculate the distance to that to make that the radius and maybe add a small padding to it (I would include that padding in the extendBackground documentation).
> @axel-op Do you have an idea of how to do that? I will add tests first, I think.
> I am fairly confident about my src/rendering implementations. Not sure how familiar you are with the rendering layer in Flutter, but you can take a look if you want.

I actually know very little about that

**Comment by creativecreatorormaybenot** [comment]
> Anyway, could you launch the example's main file and tell me if everything is fine for you?

Can you elaborate? I see no changes in here.

**Comment by creativecreatorormaybenot** [comment]
> I actually know very little about that

Well, I managed to implement it. It helps to know how the rendering layer works 😅   
However, I am having troubles implementing the widget tests. I would like to test whether an item is visible or not, but it seems that the `WidgetTester` does not care about that and will tap or "see" any widget in the tree that is not out of screen.

**Comment by axel-op** [comment]
>> Anyway, could you launch the example's main file and tell me if everything is fine for you?
> 
> Can you elaborate? I see no changes in here.

This is what I have:

**Comment by creativecreatorormaybenot** [comment]
Same, I did not run the example for some time. I will look into it. I must have changed something by accident.

**Comment by creativecreatorormaybenot** [comment]
Indeed, I forgot the clamp. Thanks for pointing it out 😃

**Comment by axel-op** [comment]
🤷‍♂️

File: lib/src/widgets/overlay.dart [hunk 14]
```diff
@@ -763,10 +763,10 @@ class _TapTarget extends StatelessWidget {
 ///    inner area, so you do not have to worry about that.
 ///  * [extendBackground] will expand the background circle if necessary.
 ///    The radius will be increased until the content fits within the circle's area
-///    and a padding of [BackgroundContentLayoutDelegate.outerContentPadding] will be added.
+///    and a padding of 6 will be added.
```
Comment (axel-op) [inline lib/src/widgets/overlay.dart]: Of 4, then

Comment (creativecreatorormaybenot) [reply lib/src/widgets/overlay.dart]: Well, I guess it was a seperate commit that fixed this 😄

**Comment by creativecreatorormaybenot** [comment]
Do you have any idea for the tests?

**Comment by axel-op** [comment]
> Do you have any idea for the tests?

Maybe you should test if a button is still accessible when the 'description' widget overflows (shouldn't be), and when it is clipped (should be).

**Comment by creativecreatorormaybenot** [comment]
Should it not be the other way round? Anyway, I mean any idea for how we could get them working. I pushed something for tests already, but it does not work.

**Comment by axel-op** [comment]
I know, and I was suggesting to do this in the following way: you place a button outside of `DescribedFeatureOverlay`, in a position where it will be covered by the 'description' widget when it overflows, and won't when it is clipped

**Comment by creativecreatorormaybenot** [comment]
I cannot get it working. I don't know if it is worth it to be honest.

Also, about the current test implementation: I think that `setSurfaceSize` is not actually required, but it was an attempt of me to get it working.

The button is definitely covered by the overlay, though, because not starting the feature discovery will make the tap pass.

**Comment by creativecreatorormaybenot** [comment]
I will call that done.

You can fix the `ignore` test if you want, but I will merge this for now.
